services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -0,0 +1,339 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.protobuf.service;
+
+import com.google.common.base.Preconditions;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.Descriptors.FileDescriptor;
+import com.google.protobuf.Descriptors.MethodDescriptor;
+import com.google.protobuf.Descriptors.ServiceDescriptor;
+
+import io.grpc.ExperimentalApi;
+import io.grpc.InternalNotifyOnServerBuild;
+import io.grpc.Server;
+import io.grpc.ServerServiceDefinition;
+import io.grpc.Status;
+import io.grpc.protobuf.ProtoFileDescriptorSupplier;
+import io.grpc.reflection.v1alpha.ErrorResponse;
+import io.grpc.reflection.v1alpha.ExtensionNumberResponse;
+import io.grpc.reflection.v1alpha.ExtensionRequest;
+import io.grpc.reflection.v1alpha.FileDescriptorResponse;
+import io.grpc.reflection.v1alpha.ListServiceResponse;
+import io.grpc.reflection.v1alpha.ServerReflectionGrpc;
+import io.grpc.reflection.v1alpha.ServerReflectionRequest;
+import io.grpc.reflection.v1alpha.ServerReflectionResponse;
+import io.grpc.reflection.v1alpha.ServiceResponse;
+import io.grpc.stub.StreamObserver;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * Provides a reflection service for Protobuf services (including the reflection service itself).
+ *
+ * <p>Throws an exception if the set of services contain multiple protobuf files with declarations
+ * of the same service, method, type, or extension.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
+public final class ProtoReflectionService extends ServerReflectionGrpc.ServerReflectionImplBase
+    implements InternalNotifyOnServerBuild {
+
+  private Server server;
+
+  /**
+   * Retrieves the services registered to the server at build time, and stores those services with a
+   * {@link ProtoFileDescriptorSupplier}. Subsequent calls to the method will have no effect.
+   */
+  @Override
+  public void notifyOnBuild(Server server) {
+    if (this.server == null) {
+      this.server = Preconditions.checkNotNull(server, "server");
+    }
+  }
+
+  @Override
+  public StreamObserver<ServerReflectionRequest> serverReflectionInfo(
```

Comment:
There are no real users of the methods on ServerCallStreamObserver, but many more for ClientCall/ServerCall. You can `git grep \\.request` to see some.

Both of those snippets didn't seem quite right though. You should `request(1)` initially, and after you are done with the previous request. That will mean the code will be processing either 0 or 1 messages at any given time.

Just something thrown together quickly:
```java
final ServerCallStreamObserver<ServerReflectionResponse> serverCallObserver =
    (ServerCallStreamObserver<ServerReflectionResponse>) responseObserver;
serverCallObserver.disableAutoInboundFlowControl();
serverCallObserver.request(1);
class MyStreamObserver implements Runnable {
  ServerReflectionRequest req;
  @Override
  public void onNext(ServerReflectionRequest r) {
    assert req == null;
    req = r;
    doWork();
  }

  @Override // as OnReadyHandler
  public void run() {
    if (req != null) {
      doWork();
    }
  }

  private void doWork() {
    while (serverCallObserver.isReady()) { // This could be a do-while loop
      ServerReflectionResponse resp = computeNextResponse();
      if (resp == null) { // I'm assuming this is the signal for "done"
        // We processed the current message, so now we request another message
        req = null;
        serverCallObserver.request(1);
        return;
      }
      serverCallObserver.onNext(resp);
      // This could also be a fine place to check if done with the current message,
      // although then the loop works slightly better as while instead of do-while
      // (not enough to be bothered by, but more from a conceptual standpoint)
    }
  }
};
MyStreamObserver mso = new MyStreamObserver();
serverCallObserver.setOnReadyHandler(mso);
return mso;
```

      
      serverCallObserver.request(1);